tests/FSharp.Compiler.ComponentTests/Language/FixedExpressionTests.fs

```diff
@@ -0,0 +1,721 @@
+                        (Warning 9, Line 10, Col 5, Line 10, Col 18, """Uses of this construct may result in the generation of unverifiable .NET IL code. This warning can be disabled using '--nowarn:9' or '#nowarn "9"'.""")
+                    ])
+            else
+                (fun comp ->
+                    comp
+                    |> shouldFail
+                    |> withDiagnostics [
+                        (Warning 9, Line 9, Col 9, Line 9, Col 12, """Uses of this construct may result in the generation of unverifiable .NET IL code. This warning can be disabled using '--nowarn:9' or '#nowarn "9"'.""")
+                        (Error 3350, Line 9, Col 9, Line 9, Col 12, """Feature 'extended fixed expressions for byrefs, inrefs, and GetPinnableReference' is not available in F# 7.0. Please use language version 'PREVIEW' or greater.""")])
+        
+    [<Theory>]
+    [<InlineData("7.0", false)>]
+    [<InlineData("preview", true)>]
+    let ``Pin type with method GetPinnableReference : unit -> inref<T>`` (langVersion, featureShouldActivate) =
+        featureShouldActivate |> ignore
+        Fsx """
+open Microsoft.FSharp.NativeInterop
+
+type RefField<'T>(_value) =
+    let mutable _value = _value 
+    member this.GetPinnableReference () : inref<'T> = &_value
+
+let pinIt (thing: RefField<'T>) =
+    use ptr = fixed thing
+    NativePtr.get ptr 0
+"""
+        |> withLangVersion langVersion
+        |> ignoreWarnings
+        |> typecheck
+        |>  if featureShouldActivate then
+                (fun comp ->
+                    comp
+                    |> shouldSucceed
+                    |> withDiagnostics [
+                        (Warning 9, Line 9, Col 9, Line 9, Col 12, """Uses of this construct may result in the generation of unverifiable .NET IL code. This warning can be disabled using '--nowarn:9' or '#nowarn "9"'.""")
+                        (Warning 9, Line 10, Col 5, Line 10, Col 18, """Uses of this construct may result in the generation of unverifiable .NET IL code. This warning can be disabled using '--nowarn:9' or '#nowarn "9"'.""")
+                    ])
+            else
+                (fun comp ->
+                    comp
+                    |> shouldFail
+                    |> withDiagnostics [
+                        (Warning 9, Line 9, Col 9, Line 9, Col 12, """Uses of this construct may result in the generation of unverifiable .NET IL code. This warning can be disabled using '--nowarn:9' or '#nowarn "9"'.""")
+                        (Error 3350, Line 9, Col 9, Line 9, Col 12, """Feature 'extended fixed expressions for byrefs, inrefs, and GetPinnableReference' is not available in F# 7.0. Please use language version 'PREVIEW' or greater.""")
+                    ])
+        
+    [<Theory>]
+    [<InlineData("7.0", false)>]
+    [<InlineData("preview", true)>]
+    let ``Pin type with extension method GetPinnableReference : unit -> byref<T>`` (langVersion, featureShouldActivate) =
```

Comment:
I think these tests would read better if split by logical branches, even for the price of some code dupe. Not a must though.

Reply:
Would it be acceptable to extract these snippets into their own files, then just write two test cases for each of them? That way we can improve readability while avoiding duplication.

Reply:
Done

Reply:
Looks much clearer, thanks!